h/accounts/test/views_test.py

```diff
@@ -829,232 +727,280 @@ def test_activate_event_when_validation_succeeds(event,
+        NotificationsController(request).notifications_form()
 
-    assert result['code'] == 401
-    assert any('pwd' in err for err in result['errors'])
 
+@notifications_fixtures
+def test_notifications_form_sets_subscriptions_data_in_form(authn_policy,
+                                                            subscriptions_model):
+    request = DummyRequest()
+    authn_policy.authenticated_userid.return_value = 'fiona'
+    subscriptions_model.get_subscriptions_for_uri.return_value = [
+        FakeSubscription('reply', True),
+        FakeSubscription('foo', False),
+    ]
+    controller = NotificationsController(request)
+    controller.form = form_validating_to({})
+
+    controller.notifications_form()
+
+    controller.form.set_appstruct.assert_called_once_with({
+        'notifications': set(['reply']),
+    })
 
-@pytest.mark.usefixtures('activation_model')
-def test_disable_user_sets_random_password(form_validator, user_model):
-    """Check if the user is disabled."""
-    request = Mock(method='POST', authenticated_userid='john')
-    form_validator.return_value = (None, {"username": "john", "pwd": "doe"})
 
-    user = FakeUser(password='abc')
-    user_model.get_by_userid.return_value = user
+@notifications_fixtures
+def test_notifications_404s_if_not_logged_in(authn_policy):
+    request = DummyRequest(post={})
+    authn_policy.authenticated_userid.return_value = None
 
-    profile = ProfileController(request)
-    profile.disable_user()
+    with pytest.raises(httpexceptions.HTTPNotFound):
+        NotificationsController(request).notifications()
 
-    assert user.password == user_model.generate_random_password.return_value
 
+@notifications_fixtures
+def test_notifications_with_invalid_data_returns_form():
+    request = DummyRequest(post={})
+    controller = NotificationsController(request)
+    controller.form = invalid_form()
 
-def test_disable_user_with_no_authenticated_user():
-    exc = ProfileController(Mock(authenticated_userid=None)).disable_user()
+    result = controller.notifications()
 
-    assert isinstance(exc, httpexceptions.HTTPUnauthorized)
+    assert 'form' in result
 
 
-@patch('h.accounts.views.Subscriptions')
-def test_unsubscribe_sets_active_to_False(Subscriptions):
-    """It sets the active field of the subscription to False."""
-    Subscriptions.get_by_id.return_value = Mock(
-        uri='acct:bob@hypothes.is', active=True)
-    request = MagicMock(
-        authenticated_userid='acct:bob@hypothes.is',
-        GET={'subscription_id': 'subscription_id'}
-    )
+@notifications_fixtures
+def test_notifications_form_with_valid_data_updates_subscriptions(authn_policy,
+                                                                  subscriptions_model):
+    request = DummyRequest(post={})
+    authn_policy.authenticated_userid.return_value = 'fiona'
+    subs = [
+        FakeSubscription('reply', True),
+        FakeSubscription('foo', False),
```

Comment:
ie. Is the subscription type required to be a value from an enum or can it be arbitrary?